### PR TITLE
Issue #12099: Add ArchUnit test to detect cycles in packages

### PR DESCRIPTION
**Pull request by Vyom-Yadav**
#12099 

Link to documentation about freezing arch rules: https://www.archunit.org/userguide/html/000_Index.html#_configuration

ViolationStore was created using property `freeze.store.default.allowStoreCreation=true`, default is `false` as we do not want it to be created in CI pipeline.

example of failure in test from diff: https://github.com/checkstyle/checkstyle/pull/12190/commits/4b61c26b93ced426126756e97a2d9454ea9a9338

### CODE REVIEW

**Comment by romani** [comment]
How it looks like if we add new cycle ? Or be able to resolve cycle?

You can create more commits in this PR to add remove/add some code to pass compilation and let arch unit to action

**Comment by romani** [comment]
@Vyom-Yadav , please resove conflict.

please create extra test commits to show how it will looks like to deal with it, we can remove such extra commits ones we confirm that all works.

**Comment by nrmancuso** [comment]
> please resolve conflict

done

> please create extra test commits to show how it will looks like to deal with it, we can remove such extra commits ones we confirm that all works.

From https://dev.azure.com/romanivanovjr/a35b6275-caac-4633-b371-79a5ab0f70c1/_apis/build/builds/11277/logs/60:

```
...
2022-09-14T03:00:55.6979513Z [ERROR]   ArchUnitTest.testClassesInApiDoNotDependOnClassesInUtil:153 api package: no classes should depend on classes that reside in any package ['com.puppycrawl.tools.checkstyle.utils', 'com.puppycrawl.tools.checkstyle.checks.javadoc.utils']
2022-09-14T03:00:55.6980195Z value of: getDetails()
2022-09-14T03:00:55.6980442Z expected to be empty
2022-09-14T03:00:55.6985944Z but was : [Method <com.puppycrawl.tools.checkstyle.api.JavadocUtil.getJavadocTags(com.puppycrawl.tools.checkstyle.api.TextBlock, com.puppycrawl.tools.checkstyle.api.JavadocUtil$JavadocTagType)> calls method <com.puppycrawl.tools.checkstyle.checks.javadoc.utils.BlockTagUtil.extractBlockTags([Ljava.lang.String;)> in (JavadocUtil.java:97), Method <com.puppycrawl.tools.checkstyle.api.JavadocUtil.getJavadocTags(com.puppycrawl.tools.checkstyle.api.TextBlock, com.puppycrawl.tools.checkstyle.api.JavadocUtil$JavadocTagType)> calls method <com.puppycrawl.tools.checkstyle.checks.javadoc.utils.InlineTagUtil.extractInlineTags([Ljava.lang.String;)> in (JavadocUtil.java:102), Method <com.puppycrawl.tools.checkstyle.api.JavadocUtil.getJavadocTags(com.puppycrawl.tools.checkstyle.api.TextBlock, com.puppycrawl.tools.checkstyle.api.JavadocUtil$JavadocTagType)> calls method <com.puppycrawl.tools.checkstyle.checks.javadoc.utils.TagInfo.getName()> in (JavadocUtil.java:116), Method <com.puppycrawl.tools.checkstyle.api.JavadocUtil.getJavadocTags(com.puppycrawl.tools.checkstyle.api.TextBlock, com.puppycrawl.tools.checkstyle.api.JavadocUtil$JavadocTagType)> calls method <com.puppycrawl.tools.checkstyle.checks.javadoc.utils.TagInfo.getName()> in (JavadocUtil.java:118), Method <com.puppycrawl.tools.checkstyle.api.JavadocUtil.getJavadocTags(com.puppycrawl.tools.checkstyle.api.TextBlock, com.puppycrawl.tools.checkstyle.api.JavadocUtil$JavadocTagType)> calls method <com.puppycrawl.tools.checkstyle.checks.javadoc.utils.TagInfo.getName()> in (JavadocUtil.java:121), Method <com.puppycrawl.tools.checkstyle.api.JavadocUtil.getJavadocTags(com.puppycrawl.tools.checkstyle.api.TextBlock, com.puppycrawl.tools.checkstyle.api.JavadocUtil$JavadocTagType)> calls method <com.puppycrawl.tools.checkstyle.checks.javadoc.utils.TagInfo.getPosition()> in (JavadocUtil.java:109), Method <com.puppycrawl.tools.checkstyle.api.JavadocUtil.getJavadocTags(com.puppycrawl.tools.checkstyle.api.TextBlock, com.puppycrawl.tools.checkstyle.api.JavadocUtil$JavadocTagType)> calls method <com.puppycrawl.tools.checkstyle.checks.javadoc.utils.TagInfo.getPosition()> in (JavadocUtil.java:114), Method <com.puppycrawl.tools.checkstyle.api.JavadocUtil.getJavadocTags(com.puppycrawl.tools.checkstyle.api.TextBlock, com.puppycrawl.tools.checkstyle.api.JavadocUtil$JavadocTagType)> calls method <com.puppycrawl.tools.checkstyle.checks.javadoc.utils.TagInfo.getValue()> in (JavadocUtil.java:118)]
2022-09-14T03:00:55.6991581Z [INFO] 
2022-09-14T03:00:55.6991882Z [ERROR] Tests run: 3859, Failures: 3, Errors: 0, Skipped: 2
2022-09-14T03:00:55.6992168Z [INFO] 
2022-09-14T03:00:55.6992813Z [INFO] ------------------------------------------------------------------------
2022-09-14T03:00:55.6993173Z [INFO] BUILD FAILURE
2022-09-14T03:00:55.6993678Z [INFO] ------------------------------------------------------------------------
2022-09-14T03:00:55.6994034Z [INFO] Total time:  03:56 min
2022-09-14T03:00:55.6994495Z [INFO] Finished at: 2022-09-14T03:00:55Z
2022-09-14T03:00:55.6995045Z [INFO] ------------------------------------------------------------------------
2022-09-14T03:00:55.6995888Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.2:test (default-test) on project checkstyle: There are test failures.
...
```

**Comment by nrmancuso** [comment]
@Vyom-Yadav do we need to add another property?
```

➜  checkstyle git:(ArchUnit-CycleDetection) ✗ mvn test -Dtest=ArchUnitCyclesCheckTest#testSlicesShouldBeFreeOfCycle test
...
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[INFO] Running com.puppycrawl.tools.checkstyle.internal.ArchUnitCyclesCheckTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 4.311 s <<< FAILURE! - in com.puppycrawl.tools.checkstyle.internal.ArchUnitCyclesCheckTest
[ERROR] testSlicesShouldBeFreeOfCycle  Time elapsed: 4.298 s  <<< ERROR!
com.tngtech.archunit.library.freeze.StoreUpdateFailedException: Updating frozen violations is disabled (enable by configuration freeze.store.default.allowStoreUpdate=true)
        at com.puppycrawl.tools.checkstyle.internal.ArchUnitCyclesCheckTest.testSlicesShouldBeFreeOfCycle(ArchUnitCyclesCheckTest.java:45)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   ArchUnitCyclesCheckTest.testSlicesShouldBeFreeOfCycle:45 » StoreUpdateFailed U...
[INFO] 
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  32.575 s
[INFO] Finished at: 2022-09-13T23:30:06-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.2:test (default-test) on project checkstyle: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/nick/CHECKSTYLE_FORKS/vyom-yadav/checkstyle/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
➜  checkstyle git:(ArchUnit-CycleDetection) ✗ git diff | wc -l
16
➜  checkstyle git:(ArchUnit-CycleDetection) ✗ git diff
diff --git a/src/test/resources/archunit.properties b/src/test/resources/archunit.properties
index 421b4e571..a5481ea27 100644
--- a/src/test/resources/archunit.properties
+++ b/src/test/resources/archunit.properties
@@ -11,3 +11,11 @@ cycles.maxNumberOfDependenciesPerEdge=550
 cycles.maxNumberToDetect=200
 
 freeze.store.default.path=config/archunit-store
+
+# must be set to true to allow the creation of a new violation store
+# default is false
+freeze.store.default.allowStoreCreation=false
+
+# can be set to false to forbid updates of the violations stored for frozen rules
+# default is true
+freeze.store.default.allowStoreUpdate=false
➜  checkstyle git:(ArchUnit-CycleDetection) ✗ 



```

**Comment by romani** [comment]
what was code change ? that caused this error

**Comment by nrmancuso** [comment]
> what was code change ? that caused this error what is required is we need to reduce or extend known store of suppressions ?

See above comment, I just moved utils to api package to resolve a bunch of violations at one time. The behavior of this freezing rule is strange, see https://github.com/checkstyle/checkstyle/pull/12190/commits/02e822ef3c778c9f5f266bcd54e07dcac9da5fd6 where I simply added one violation and have a diff of over 1,000 lines in freeze file.

**Comment by nrmancuso** [comment]
> last removal in your diff
> 
> > Method <com.puppycrawl.tools.checkstyle.xpath.XpathQueryGenerator.isMatchingByLineAndColumnAndTokenType(com.puppycrawl.tools.checkstyle.api.DetailAST)> calls method <com.puppycrawl.tools.checkstyle.api.DetailAST.getType()> in (XpathQueryGenerator.java:333)
> 
> It is not a cycle, so it might be problem with original generation of file.

This looks like all the tool did was remove the newline at the end of the file, unless I am missing soimething.

**Comment by nrmancuso** [comment]
Example of removing one violation: https://github.com/checkstyle/checkstyle/pull/12190/commits/ab8dfb89f62f243ecf1c587a4b2d9d80a4d7acdd

Diff is removal of over 1,000 lines; also test is still failing after freeze store update.

**Comment by Vyom-Yadav** [comment]
We are enabling re-freeze to minimize diff in the store in case a violation is added or removed. This would be more convenient for the developer. 

- Violation is removed: https://github.com/checkstyle/checkstyle/pull/12205
- Violation is added: https://github.com/checkstyle/checkstyle/pull/12206

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (nrmancuso) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:35]: Please do this: imagine you are a new contributor, and execution of this test has generated a new suppression. Then, in the issue, place links and information about the concept of this test and links to freeze documentation. After that, place a Javadoc here that explains how this test works, and a link to your comment in the issue.

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Item:

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:35]: Yes, it good idea to read details in actual test class. Hints on how to interpretate is also good.

What is the process if we ok to allow extra cycle?

Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:48]: @nrmancuso Please take a look at the javadoc.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:48]: does it mean that test will always pass and just update freeze file, so will enforce no cycles by simple PR review and noticing a diff - I am ok.

If user forget to place freeze file to commit, CI will fail as it will detect  a git-diff.

Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:48]: Yes, for any new violation, the store will be automatically updated, if the user commits the changes in the store, they will be seen and reviewed in the PR.

> If user forget to place freeze file to commit, CI will fail as it will detect a git-diff.

Yes

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:48]: please mention that change in freeze file need to be placed in commit.
Any change in freeze file should be committed and shared in PR.

Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:49]: > Whenever a new violation is introduced, the test will pass successfully but the frozen violations will be updated. In that case, developers should check if the new violation is expected or not. In both cases create a separate commit in the PR for discussing changes in the violation store.

Added this para to the javadoc.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:41]: Whenever new cycle is introduced or removed

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:42]: violations store will be updated, It is highly recommended to avoid new cycles and, and it is preferred to remove them. In both cases store file should be placed to commit for review by maintainers.

**Review by romani — CHANGES_REQUESTED** [review]
items

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:41]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:42]: Done.

**Review by romani — APPROVED** [review]
Ok to merge

**Comment by romani** [comment]
@Vyom-Yadav , please resolve conflict

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:43]: In both cases put update in commit for discussion

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:43]: What about:
In both cases update the commit for discussing changes in the violation store.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:43]: 'Update commit' for sounds strange .
We need some sort on "include in commit"

Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:43]: The suppressions generated by this tool are like the metadata generator; most contributors will never know/care about such changes and should commit all changes in one commit, and it is doubtful they will even see this comment until a reviewer points it out. It is responsibility of reviewer to catch extension of this file and question why we extended cycles.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:43]: Yes, lets agree on some reasonable wording, and let's merge this PR

Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:43]: In both cases commit the changes for further discussion.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:43]: Good

Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java]: Done.

**Review by romani — APPROVED** [review]
Ok to merge

**Review by nrmancuso — APPROVED** [review]
I am good to merge once CI passes

**Comment by Vyom-Yadav** [comment]
> I am good to merge once CI passes

checker-framework will fail due to #12210

File: config/archunit-store/stored.rules
Comment (rnveach) [inline config/archunit-store/stored.rules:3]: > 7999d2d1

Question: Are we required to use such naming?

**Comment by romani** [comment]
We can not merge .. red CI will affect all

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:36]: `Tests that there are no cycles in the package` => `Tests that have no cycles in the package`

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:42]: Question: I have not been part of the conversation but CI will fail on unstaged changes after run, right?

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:48]: `testSlicesShouldBeFreeOfCycle` => `testSlicesShouldBeFreeOfCycles`

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:53]: `slicesShouldBeFreeOfCycle` => `slicesShouldBeFreeOfCycles`

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:42]: Yes

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:36]: Why? Isn't `Tests that there are no cycles in the package dependencies` correct?

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:48]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:53]: Done.

File: config/archunit-store/stored.rules
Comment (Vyom-Yadav) [reply config/archunit-store/stored.rules]: It was automatically generated by ArchUnit, didn't try changing it. Changed now, and it works perfectly fine.

File: config/archunit-store/stored.rules
Comment (rnveach) [inline config/archunit-store/stored.rules:2]: Sorry one more thing I just noticed, do we really need this date? We won't be updating it right?

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:36]: "that there" sounds very odd and I would say it is not proper. "That there" is more like you are pointing to a specific thing (noun) like "that there doggy in the window" but you don't really follow up with a noun to point to. The replacement text sounds more natural to me. "Tests which have" would also be acceptable.

https://brians.wsu.edu/2016/05/25/this-here-that-there/

Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:36]: `Tests that have no`, in this `Test` is acting as a noun, similar to `Boys that have no`, whereas we should be using `Test` as a verb.

File: config/archunit-store/stored.rules
Comment (Vyom-Yadav) [reply config/archunit-store/stored.rules]: No, I don't think so, these are comments most proly, so I removed them.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:36]: Your original sentence is not using Test as a verb. It is also a noun.

Please suggest a sentence if you don't want to use my wording, but "that there" still has to change.

Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java:36]: Nick suggested `Detects cycles in package dependencies`, is that fine?

Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitCyclesCheckTest.java]: Done.

**Review by rnveach — APPROVED** [review]

**Pull request merged**
